exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/InsertStatement.kt

```diff
@@ -52,99 +52,122 @@ open class InsertStatement<Key : Any>(
+                    // TooGenericExceptionCaught suppress also can be removed
+
+                    val preparedSql = prepareSQL(TransactionManager.current(), prepared = true)
+
+                    val returnedColumnsString = columnIndexesInResultSet
+                        .mapIndexed { index, pair -> "column: ${pair.first.name}, index: ${pair.second} (columns-list-index: $index)" }
+                        .joinToString(prefix = "[", postfix = "]", separator = ", ")
+
+                    exposedLogger.error(
+                        "ArrayIndexOutOfBoundsException on processResults. " +
+                            "Table: ${table.tableName}, firstAutoIncColumn: ${firstAutoIncColumn?.name}, " +
+                            "inserted: $inserted, returnedColumnsString: $returnedColumnsString. " +
+                            "Failed SQL: $preparedSql",
+                        cause
+                    )
+                    throw cause
                 }
             }
-        }
 
-        arguments!!.forEachIndexed { itemIndx, pairs ->
-            val map = autoGeneratedKeys.getOrNull(itemIndx) ?: hashMapOf<Column<*>, Any?>().apply {
-                autoGeneratedKeys.add(itemIndx, this)
-            }
-            pairs.forEach { (col, value) ->
-                if (value != DefaultValueMarker) {
-                    val unwrappedValue = value?.let { (col.columnType as? ColumnWithTransform<Any, Any>)?.unwrapRecursive(it) } ?: value
-                    if (isColumnValuePreferredFromResultSet(col, value)) {
-                        map.getOrPut(col) { unwrappedValue }
-                    } else {
-                        map[col] = unwrappedValue
+            if (inserted > 1 && firstAutoIncColumn != null && resultSetsValues.isNotEmpty() && !currentDialect.supportsMultipleGeneratedKeys) {
+                // H2/SQLite only returns one last generated key...
+                (resultSetsValues[0][firstAutoIncColumn] as? Number)?.toLong()?.let {
+                    var id = it
+
+                    while (resultSetsValues.size < inserted) {
+                        id -= 1
+                        resultSetsValues.add(0, mutableMapOf(firstAutoIncColumn to id))
                     }
                 }
             }
+
+            assert(
+                isIgnore || resultSetsValues.isEmpty() || resultSetsValues.size == inserted ||
+                    currentDialect.supportsTernaryAffectedRowValues
+            ) {
+                "Number of autoincs (${resultSetsValues.size}) doesn't match number of batch entries ($inserted)"
+            }
         }
-        @Suppress("UNCHECKED_CAST")
-        return autoGeneratedKeys.map { ResultRow.createAndFillValues(it as Map<Expression<*>, Any?>) }
+
+        return resultSetsValues
     }
 
-    protected open fun isColumnValuePreferredFromResultSet(column: Column<*>, value: Any?): Boolean {
-        return column.columnType.isAutoInc || value is NextVal<*>
```

Comment:
Sorry if I missed the reason, but why are we dropping this in 1 go instead of deprecating it the same way as `valuesAndDefaults()` below?

Reply:
The only reason is that it was introduced quite recently and should not be used by many users. But I agree, it would be more consistent to mark it as deprecated also.